pkg/xds/experimental/client/listener.go

```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+package client
+
+import (
+	"log/slog"
+
+	"github.com/cilium/cilium/pkg/envoy/xds"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// listener is a helper structure for watchers focused on handling a single, registered callback.
+type listener struct {
+	ID      uint64
+	typeUrl string
+	cb      WatcherCallback
+	trigger chan uint64
+
+	log       *slog.Logger
+	resources xds.ResourceSource
+
+	lastVersion uint64
+	mux         lock.Mutex
+	nextVersion uint64
+}
+
+var _ xds.ResourceVersionObserver = (*listener)(nil)
+
+// HandleNewResourceVersion implements xds.ResourceVersionObserver.
+// It triggers the callback process when a new version appears.
+func (l *listener) HandleNewResourceVersion(typeUrl string, version uint64) {
+	if typeUrl != l.typeUrl {
+		return
+	}
+	l.mux.Lock()
+	defer l.mux.Unlock()
+	if version <= l.lastVersion {
+		return
+	}
+	l.nextVersion = version
+	select {
+	case l.trigger <- version:
+	default:
+	}
+}
+
+// process waits for the listener trigger (new resource version) and invokes the callback function.
```

Comment:
The channel and goroutine interlock warrants some discussion.

Reply:
Please elaborate

Reply:
process() is called from a go routine and for example if the parent ctx is canceled we should exit

Reply:
made the code simpler